DashAI/back/models/hugging_face/vit_transformer.py

```diff
@@ -0,0 +1,164 @@
+
+        Returns
+        -------
+        Function
+            a function that preprocesses images and outputs a dictionary
+            containing processed images and corresponding labels.
+        """
+
+        def preprocess_images(examples):
+            inputs = self.feature_extractor(
+                images=examples[input_column], return_tensors="pt", size=224
+            )
+            inputs["labels"] = examples[output_column]
+            return inputs
+
+        return preprocess_images
+
+    def fit(self, dataset: DashAIDataset):
+        """Fine-tuning the pre-trained model
+
+        Parameters
+        ----------
+        dataset : DashAIDataset
+            DashAiDataset with training data
+
+        """
+
+        input_column = dataset.inputs_columns[0]
+        output_column = dataset.outputs_columns[0]
+
+        feature_extractor_func = self.get_preprocess_images(input_column, output_column)
+        dataset = dataset.map(feature_extractor_func, batched=True)
+
+        # Arguments for fine-tuning
+        training_args = TrainingArguments(
+            output_dir="DashAI/back/user_models/temp_checkpoints_vit",
+            save_steps=1,
+            save_total_limit=1,
+            **self.training_args,
+        )
+
+        # The Trainer class is used for fine-tuning the model.
+        trainer = Trainer(
+            model=self.model,
+            args=training_args,
+            train_dataset=dataset,
+        )
+
+        trainer.train()
+        self.fitted = True
+        shutil.rmtree(
+            "DashAI/back/user_models/temp_checkpoints_vit", ignore_errors=True
+        )
+        return
+
+    def predict(self, dataset: DashAIDataset):
+        """
+        Make a prediction with the fine-tuned model
+
+        Parameters
+        ----------
+        dataset : DashAIDataset
+            DashAIDataset with image data
+
+        Returns
+        -------
+        Numpy Array
+            Numpy array with the probabilities for each class
+        """
+        if not self.fitted:
+            raise NotFittedError(
+                f"This {self.__class__.__name__} instance is not fitted yet. Call 'fit'"
+                " with appropriate arguments before using this estimator."
+            )
+
+        input_column = dataset.inputs_columns[0]
+        output_column = dataset.outputs_columns[0]
+        preprocess_images = self.get_preprocess_images(input_column, output_column)
+
+        dataset = dataset.map(preprocess_images, batched=True)
+        dataset.set_format("torch", columns=["pixel_values", "labels"])
+
+        probabilities = []
+
+        # Iterate over each batch in the dataset
+        for i in range(len(dataset)):
+            # Prepare a batch of images for the model
+            batch = dataset[i]
+
+            # Make sure that the tensors are in the correct device.
+            batch = {k: v.to(self.model.device) for k, v in batch.items()}
```

Comment:
Just so we think about it with the team later, what would happen if my device fails to move the tensors to the gpu and throws an exception. What would happen to the front end?

Reply:
It's a good question and we should talk about it.